File: examples/testlib/art/cartoon/stickman/feet.js
const defaultFeet = {
  length: 2.2,
  stroke: "#f97316",
  strokeWidth: 0.5
};

const helperCollection = typeof helpers === "object" ? helpers : null;
const normalizeInput = helperCollection?.normalizeInput;
const normalizePoint = helperCollection?.normalizePoint;
const resolveNumber = helperCollection?.resolveNumber;

function requireHelper(fn, name) {
  if (typeof fn !== "function") {
    throw new Error(`cartoon/helpers/${name}.js must export a function as helpers.${name}`);
  }
}

requireHelper(normalizeInput, "normalizeInput");
requireHelper(normalizePoint, "normalizePoint");
requireHelper(resolveNumber, "resolveNumber");

function clampPositive(value, fallback, min = 0.01) {
  const result = resolveNumber(value, fallback);
  return result >= min ? result : fallback;
}

function normalizeSide(value, fallback = "left") {
  if (typeof value === "string") {
    const lowered = value.trim().toLowerCase();
    if (lowered === "right" || lowered === "left") {
      return lowered;
    }
  }
  return fallback;
}

function normalizeDirection(value, fallback = "left") {
  if (typeof value === "string") {
    const lowered = value.trim().toLowerCase();
    if (lowered === "right" || lowered === "left") {
      return lowered;
    }
    if (lowered === "center" || lowered === "middle") {
      return "center";
    }
  }
  return fallback;
}

function createFeet(rawOptions = {}) {
  const options = normalizeInput(rawOptions, {});
  const anklePoint = normalizePoint(options.anklePoint ?? options.anchor ?? options.position, [0, 0]);
  const side = normalizeSide(options.side);
  const directionHint = normalizeDirection(options.directionHint, normalizeDirection(side, "left"));
  const lineLength = clampPositive(options.length, defaultFeet.length, 0.05);
  const style = normalizeInput(options.style, {});
  const stroke =
    typeof options.stroke === "string"
      ? options.stroke
      : typeof style.stroke === "string"
        ? style.stroke
        : defaultFeet.stroke;
  const strokeWidth = clampPositive(style.width ?? options.strokeWidth, defaultFeet.strokeWidth, 0.01);
  const half = lineLength / 2;
  const sign = directionHint === "right" ? 1 : -1;
  const fromPoint =
    directionHint === "center" ? [anklePoint[0] - half, anklePoint[1]] : anklePoint;
  const toePoint =
    directionHint === "center"
      ? [anklePoint[0] + half, anklePoint[1]]
      : [anklePoint[0] + sign * lineLength, anklePoint[1]];
  const centerPoint = directionHint === "center" ? anklePoint : toePoint;

  return {
    graphics: [
      {
        type: "line",
        from: fromPoint,
        to: toePoint,
        stroke,
        width: strokeWidth
      }
    ],
    anklePoint,
    center: centerPoint,
    toePoint,
    side,
    direction: directionHint,
    length: lineLength
  };
}

return createFeet;

File: examples/testlib/art/cartoon/stickman/feet.fs
(options)=>
{
  defaults:{ length:2.2; stroke:"#f97316"; strokeWidth:0.5 };

  input:options ?? {};
  anklePoint:if input.anklePoint != null then input.anklePoint else if input.anchor != null then input.anchor else if input.position != null then input.position else [0,0];
  side:normalizeSide(input.side, "left");
  directionHint:normalizeDirection(input.directionHint, normalizeDirection(side, "left"));
  lineLength:clampPositive(input.length, defaults.length, 0.05);
  style:input.style ?? {};
  stroke:input.stroke ?? style.stroke ?? defaults.stroke;
  strokeWidth:clampPositive(style.width ?? input.strokeWidth, defaults.strokeWidth, 0.01);
  half:lineLength / 2;
  sign:if directionHint = "right" then 1 else -1;
  fromPoint:if directionHint = "center" then [anklePoint[0] - half, anklePoint[1]] else anklePoint;
  toePoint:if directionHint = "center" then [anklePoint[0] + half, anklePoint[1]] else [anklePoint[0] + sign * lineLength, anklePoint[1]];
  centerPoint:if directionHint = "center" then anklePoint else toePoint;

  eval {
    graphics:[
      {
        type:"line";
        from:fromPoint;
        to:toePoint;
        stroke:stroke;
        width:strokeWidth;
      }
    ];
    anklePoint:anklePoint;
    center:centerPoint;
    toePoint:toePoint;
    side:side;
    direction:directionHint;
    length:lineLength;
  };

  clampPositive:(value, fallback, min)=> {
    minValue:min ?? 0.01;
    resolved:if value = null then fallback else value;
    return if resolved >= minValue then resolved else fallback;
  };

  normalizeSide:(value, fallback)=> {
    defaultSide:fallback ?? "left";
    eval if value = null then defaultSide else {
      lowered:text.lower(format(value));
      eval if lowered = "right" then "right" else if lowered = "left" then "left" else defaultSide;
    }
  };

  normalizeDirection:(value, fallback)=> {
    defaultDirection:fallback ?? "left";
    eval if value = null then defaultDirection else {
      lowered:text.lower(format(value));
      eval if lowered = "center" then "center"
      else if lowered = "middle" then "center"
      else if lowered = "right" then "right"
      else if lowered = "left" then "left"
      else defaultDirection;
    }
  };
}
